src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,565 @@
+    this->proximalDecrement           = proximalDecrement;
+    this->proximalSynapseThreshold    = proximalSynapseThreshold;
+    this->stability_rate              = stability_rate;
+    this->fatigue_rate                = fatigue_rate;
+
+    SDR proximalInputs(  proximalInputDimensions );
+    SDR inhibitionAreas( inhibitionDimensions );
+    cellDimensions_ = inhibitionAreas.dimensions;
+    cellDimensions_.push_back( cellsPerInhibitionArea );
+    SDR cells( cellDimensions_ );
+    rng_ = Random(seed);
+
+    // Setup the proximal segments & synapses.
+    proximalConnections.initialize(cells.size, proximalSynapseThreshold);
+    SDR_Sparsity PP_Sp(            proximalInputs.dimensions, 10 * proximalInputs.size);
+    SDR_ActivationFrequency PP_AF( proximalInputs.dimensions, 10 * proximalInputs.size);
+    UInt cell = 0u;
+    for(auto inhib = 0u; inhib < inhibitionAreas.size; ++inhib) {
+      inhibitionAreas.setSparse(SDR_sparse_t{ inhib });
+      for(auto c = 0u; c < cellsPerInhibitionArea; ++c, ++cell) {
+        for(auto s = 0u; s < proximalSegments; ++s) {
+          auto segment = proximalConnections.createSegment( cell );
+
+          // Make synapses.
+          potentialPool( inhibitionAreas, proximalInputs );
+          for(const auto presyn : proximalInputs.getSparse() ) {
+            auto permanence = initProximalPermanence();
+            proximalConnections.createSynapse( segment, presyn, permanence);
+          }
+          proximalConnections.raisePermanencesToThreshold( segment,
+                          proximalSynapseThreshold, proximalSegmentThreshold );
+          PP_Sp.addData( proximalInputs );
+          PP_AF.addData( proximalInputs );
+        }
+      }
+    }
+    tieBreaker_.resize( proximalConnections.numSegments() );
+    for(auto i = 0u; i < tieBreaker_.size(); ++i) {
+      tieBreaker_[i] = 0.01f * rng_.getReal64();
+    }
+    proximalMaxSegment_.resize( cells.size );
+    AF = new SDR_ActivationFrequency( {cells.size, proximalSegments}, period );
+    AF->initializeToValue( sparsity / proximalSegments );
+
+    // Setup the distal dendrites
+    distalConnections.initialize(
+        /* columnDimensions */            {2048},
+        /* cellsPerColumn */              1,
+        /* activationThreshold */         distalSegmentThreshold,
+        /* initialPermanence */           0.21, // TODO: Needs to be a parameter!!!
+        /* connectedPermanence */         distalSynapseThreshold,
+        /* minThreshold */                distalSegmentMatch,
+        /* maxNewSynapseCount */          distalAddSynapses,
+        /* permanenceIncrement */         distalIncrement,
+        /* permanenceDecrement */         distalDecrement,
+        /* predictedSegmentDecrement */   distalMispredictDecrement,
+        /* seed */                        rng_(),
+        /* maxSegmentsPerCell */          distalMaxSegments,
+        /* maxSynapsesPerSegment */       distalMaxSynapsesPerSegment,
```

Comment:
actually, instead of having to specify all of these in CP's constructor, I'd suggest to optionally pass `TemporalMemory *tm`
- some tasks (as MNIST) do not need the distal connections/temporal mem. 
- in those that do, you need to instantialized tm beforehand. So you can pass the &tm aswell.

Reply:
Instead I will provide default parameters which have the TM disabled, and methods which ignore & pass zero'd input to the TM.

Reply:
>  and methods which ignore & pass zero'd input to the TM. 

add some if to skip it instead, it'd be wasteful to call TM if unneeded

Reply:
Using a TM which has been disabled by construction should be very fast & use very little memory.  It will have 0 segments & 0 synapses.  It amounts to iterating through some empty lists, and doing some book-keeping for each cell.  

I'd rather keep it all intact because it simplifies the implementation at no cost.

Reply:
>  It will have 0 segments & 0 synapses. It amounts to iterating through some empty lists, and doing some book-keeping for each cell. 

this still sounds suboptimal to me, and we'll have trouble with requiring TM with dumbed-down params. 
Let's see how the code looks. 
What I suggest is: 

```
const int TM_DISABLED = -1;
Params default = {..., distalMaxSegments = TM_DISABLED, ...};

...
if(distalMaxSegments == TM_DISABLED) continue; 
//else do distal learning on TM outputs
```

Reply:
> suboptimal

And if I made it optimal it would be premature optimization.  I want to keep this code simple where possible.  Later, once this code is verified working, then we should profile it on a variety of use cases including one with no distal dendrites.